Add watch mode to php unit tests

eduardoumpierre opened the pull request:
Closes #677 

#### Changes proposed in this Pull Request

- Add an NPM script to listen to PHP unit tests.

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->
- Checkout to this branch `add/677-unit-test-watcher`
-  Run `composer install`
* **Valet configuration**
  * Add the following code to your `.vscode/tasks.json`
  ```json
  {
      "label": "Run Tests - Watch Mode",
      "type": "shell",
      "command": "TMPDIR=~/tmp vendor/bin/phpunit-watcher watch --configuration phpunit.xml.dist",
      "group": "test",
      "presentation": {
        "showReuseMessage": false,
      },
      "problemMatcher": []
  },
  {
      "label": "Run Current Test - Watch Mode",
      "type": "shell",
      "command": "TMPDIR=~/tmp vendor/bin/phpunit-watcher watch --configuration phpunit.xml.dist --filter ${selectedText}",
      "group": "test",
      "presentation": {
        "showReuseMessage": false,
      },
      "problemMatcher": []
  }
  ```
  - On vscode, press `⌘ + shift + p`, select `Tasks: Run Task`, and click on `Run Tests - Watch mode` 
    - Or select a test class name or test method name and select `Run Current Test - Watch mode`
  - It should run all the tests, and then wait for input listening to code changes
- **Docker setup** 
  - Run `npm run up:recreate`
  - Run `npm run test:php-watch`
    - Or run `npm run test:php-watch -- --filter <test_name>`
  - It should run all the tests, and then wait for input listening to code changes

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

eduardoumpierre committed:
Add phpunit-watcher composer dependency

eduardoumpierre committed:
Add watch flag to run-tests script

phpunit-watcher is used instead of regular phpunit
in case of passing -w flag as a parameter.

New test:php-watch npm script added to package.json

eduardoumpierre committed:
Fix PHP unit tests watcher script

Update tests documentation to include new npm script.

eduardoumpierre committed:
Add changelog entry

allie500 reviewed (approved):
Thanks for working on this. The code changes are straightforward and make sense. 

Both of these cases worked for me on Docker:

`npm run test:php-watch` - ran all tests.

`npm run test:php-watch -- --filter <test_name>` - ran the specific test I passed to it.


I also tested the options provided by the watcher at the end of a test run:

```
Press a to run all tests.
Press t to filter by test name.
Press p to filter by file name.
Press g to filter by group name.
Press s to filter by test suite name.
Press r to run tests with a random seed.
Press q to quit the watcher.
Press Enter to trigger a test run.
```

Run all tests (a) works.

```
Time: 00:50.347, Memory: 171.00 MB

OK, but incomplete, skipped, or risky tests!
Tests: 1294, Assertions: 2983, Skipped: 1.

Starting PHPUnit
```

Filter by test name (t) works.

```
Time: 00:00.220, Memory: 93.00 MB

OK (1 test, 2 assertions)

Starting PHPUnit with arguments: `--filter=test_it_should_not_update_available_currencies_when_enabled_payment_methods_do_not_need_it`
```

Filter by file name (p) didn't work on my test when passing the relative path. 

```
Class 'test-class-wc-payment-gateway-wcpay-process-payment' could not be found in '/var/www/html/wp-content/plugins/woocommerce-payments/tests/unit/test-class-wc-payment-gateway-wcpay-process-payment.php'.

Starting PHPUnit with arguments: `tests/unit/test-class-wc-payment-gateway-wcpay-process-payment.php`
```

The actual class name in that file is: `WC_Payment_Gateway_WCPay_Process_Payment_Test`.


Filter by group name (g) worked on the only group I could find in the tests: `frontend-tests` in `tests/unit/multi-currency/test-class-frontend-currencies.php:17`.

```
Time: 00:04.061, Memory: 95.00 MB

OK (35 tests, 39 assertions)

Starting PHPUnit with arguments: `--group=frontend-tests`
```

Filter by test suite name (s) works with both test-suites currently configured in the `phpunit.xml.dist` file.

```
Time: 00:40.080, Memory: 157.00 MB

OK, but incomplete, skipped, or risky tests!
Tests: 905, Assertions: 2402, Skipped: 1.

Starting PHPUnit with arguments: `--testsuite=WCPay`
```

```
Time: 00:10.543, Memory: 99.00 MB

OK (389 tests, 581 assertions)

Starting PHPUnit with arguments: `--testsuite=WCPayMultiCurrency`
```

Running the tests with a random seed (r) worked to run them in random order but some fail as a result. I guess that is probably to be expected.

I think this PR is good to merge.

eduardoumpierre committed:
Increase phpunit-watcher timeout and disable phpunit-watcher notifications

Currently using 600 (10 minutes) as it depends on the developer's machine.
This value can be changed as necessary.

eduardoumpierre committed:
Fix composer.lock

jessepearson reviewed (approved):
Works for me in Valet once we were able to figure out what the task config for VSCode should be. Thanks for adding this 👍
